src/hooks/useMediaRecorder.js

```diff
@@ -1,19 +1,66 @@
-import { useEffect, useState } from 'react'
-
-export default function useMediaRecorder (stream, trigger, events) {
-    const [mediaRecorder, setMediaRecorder] = useState(null)
-
-    useEffect(() => {
-        if( mediaRecorder === null ) {
-            let mr = null
-            mr = new MediaRecorder(stream)
-            mr.onstop = events.onStop
-            mr.onstart = events.onStart
-            mr.ondataavailable = events.ondataavailable
-            setMediaRecorder(mr)
-        }
-        // eslint-disable-next-line react-hooks/exhaustive-deps
-    }, [trigger])
-
-    return mediaRecorder
-}
\ No newline at end of file
+import {useMemo, useState} from "react";
+
+/**
+ * This media recorder hook keeps all the logic needed to perform an audio recording decoupling it from any component
+ * so it can be reused if needed.
+ *
+ * This way, when it comes to testing you only need to mock this `recorder` hook and its returned values.
+ *
+ * The hook in use would look like this:
+ * const { recording, recordings, setRecordings, isRecording } = useMediaRecorded(stream);
+ *
+ * The `setRecordings` method is exposed so that we can perform modifications to the list of recordings from the component
+ * avoiding to bloat the logic of this hook (tough it could also be a good idea to do so) but we kept it like this for simplicity
+ *
```

Comment:
No, I like what you've done here. The single responsibility of this hook should be to generate a recording as a blob object and "return" it (or update state or whatever). If we want to manipulate the list of recordings, that can be easily done elsewhere, I think.

Reply:
The more I read this the more I like what you've done here. It's really perfect and elegantly solves the problems I was struggling the most with (had a stream already been started? What do I put in the onstop handler? Where do I store recordings?) Really well done. Thank you.